src/site/_includes/components/tags/Image.js

```diff
@@ -0,0 +1,211 @@
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * Adds two new shortcodes to Eleventy: Image and Figure.
+ * These shortcodes will convert local image paths to image CDN paths depending
+ * on the build environment. This allows authors to dev with local images but
+ * switch to using image CDN images with srcset and other fanciness during
+ * production.
+ *
+ * We use the custom tags API instead of Eleventy's universal shortcodes here
+ * because we need access to the page context object (specifically the URL for
+ * the page that's being rendered). Unfortunately Eleventy's universal
+ * shortcodes don't have access to this information, so custom tags it is!
+ *
+ * Example usage:
+ *
+ * {%
+ *  Image
+ *  src="./label-a11y-lint.png",
+ *  alt="hello",
+ *  maxWidth="300px"
+ * %}
+ *
+ * {%
+ *  Figure
+ *  src="./label-a11y-lint.png",
+ *  alt="hello",
+ *  type="fullbleed screenshot",
+ *  maxWidth="300px"
+ * %}
+ *  Figcaption text goes here! Can be `markdown`.
+ * {% endFigure %}
+ *
+ */
+
+const path = require('path');
+const {oneLine} = require('common-tags');
+const {ifDefined} = require('../helpers');
+const md = require('markdown-it')();
+const stripLanguage = require('../../../_filters/strip-language');
+const imageCdn = require('../../../_data/site').imageCdn;
+
+/* eslint-disable no-invalid-this, max-len */
+
+// -----------------------------------------------------------------------------
+// IMAGE
+// -----------------------------------------------------------------------------
+
+/**
+ * Takes a path to an image and converts it to an image CDN path if we're in
+ * a production environment.
+ * @param {!string} src A path to an image asset.
+ * @param {!Object} ctx An eleventy context object for the current page.
+ * @return {string} An image path. May be converted to an image CDN path if
+ * it's a production environment.
+ */
+function getImagePath(src, ctx) {
+  let imagePath = src;
+  if (process.env.ELEVENTY_ENV === 'prod') {
+    imagePath = path.join(stripLanguage(ctx.page.url), src);
+    imagePath = new URL(imagePath, imageCdn).href;
+  }
+  return imagePath;
+}
+
+/**
+ * Render an image element as an HTML string.
+ * @param {!Object} args An args object containing src and optional alt and
+ * maxWidth.
+ * @param {!Object} ctx An eleventy context object.
+ * @return {string}
+ */
+function renderImage({src, alt, maxWidth}, ctx) {
+  const imagePath = getImagePath(src, ctx);
+  let style;
+  if (maxWidth) {
+    style = `max-width: ${maxWidth};`;
+  }
+  // Using oneLine seems to make Nunjucks' SafeString function happy.
```

Comment:
is it handling escaping of `${imagePath}` and etc? If I provide an image with "'s is that fine?

Reply:
The `oneLine` method just trims all the whitespace and newlines so the text appears as a single line.

The `SafeString` method doesn't seem to be well documented but it looks like if there is anything fishy with the HTML it just escapes the whole thing, turning brackets into `&lt;` and `&gt;`. I found that if my HTML string has newlines in it, it seems to consider that suspect and escapes the whole thing.